services/webnn/ort/graph_builder_ort.cc

```diff
@@ -865,30 +941,34 @@ void GraphBuilderOrt::AddElementWiseLogicalOperation(
         absl::get<const mojom::ElementWiseUnary*>(operation);
     lhs = GetOperandNameById(element_wise_unary->input_operand_id);
 
-    // Some ONNX logical operators only support bool input.
+    // For some ONNX logical operators that require bool input, add a cast if
+    // the operand is not in the skipped list.
     if (element_wise_unary->kind ==
-        mojom::ElementWiseUnary::Kind::kLogicalNot) {
+            mojom::ElementWiseUnary::Kind::kLogicalNot &&
+        !bool_operands_.count(element_wise_unary->input_operand_id)) {
       lhs = PrependCast(lhs, ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL);
     }
 
     inputs = {lhs.c_str()};
   }
 
-  const std::string bool_output = GenerateNextOperandName();
-  std::array<const char*, 1> outputs = {bool_output.c_str()};
-
-  model_editor_.AddNode(op_type, node, inputs, outputs);
-
   // ONNX logical operators only support bool output. To support output with the
-  // WebNN data type, it is necessary to insert a cast operator after a logical
-  // operator.
+  // WebNN data type, append a cast if the operand is not in the skipped list.
   uint64_t output_operand_id = absl::visit(
       [](const auto* op) { return op->output_operand_id; }, operation);
-  OperandDataType output_data_type =
-      GetOperand(output_operand_id).descriptor.data_type();
   std::string output = GetOperandNameById(output_operand_id);
-  AppendCast(bool_output, output,
-             OperandTypeToONNXTensorElementDataType(output_data_type));
+  if (bool_operands_.count(output_operand_id)) {
```

Comment:
I concerned that if the bool operand (which should be WebNN uint8 operand) is also another operation's input (not solely a logical operation's input), if you don't append the cast to convert back to uint8, there will be error during shape inference check. Am I right? Maybe we need a sub graph test to verify. @miaobin @huningxin

Reply:
Yes, [issue](https://github.com/shiyi9801/chromium/issues/217#issue-2979066335) has been filed and I'm working on it.

Reply:
https://github.com/shiyi9801/chromium/issues/217

Reply:
Oh, got it, thanks! @miaobin @huningxin